File: crates/phoenix-codegen/tests/roundtrip/python/harness.py
# Schema-agnostic round-trip boilerplate shared with the per-schema driver.py.
#
# driver.py holds the parts coupled to a specific generated package (the Stub
# implementing the generated Handlers Protocol, the `invoke` dispatch over the
# generated client, and the FastAPI/ASGI run loop). Everything here is generic:
# the target constant, the DriverError exception, the value/key normalization
# helpers, and the assertions that compare a contract case against an observed
# result purely by duck typing (`.model_dump()`, `.value`, `.status`, etc.) — so
# this module imports neither the `generated` package nor pydantic.

from __future__ import annotations

import json
from typing import Any

import httpx

TARGET = "python"


class DriverError(Exception):
    """A contract assertion failed."""


def enum_value(v: Any) -> Any:
    """The status field is a PostStatus enum; expose its string value so it can
    be compared numerically/stringly against the contract."""
    return v.value if hasattr(v, "value") else v


def to_snake(key: str) -> str:
    out = []
    for ch in key:
        if ch.isupper():
            out.append("_")
            out.append(ch.lower())
        else:
            out.append(ch)
    return "".join(out)


def normalize(value: Any) -> Any:
    """Recursively snake_case dict keys and coerce enums/bools so the contract's
    camelCase expectations compare equal to the Python generator's snake_case
    wire format. Numbers are left as-is and compared numerically below. Callers
    that compare two structures apply this to BOTH sides, so the key rewrite
    stays symmetric even for data-map keys this can't distinguish from fields."""
    if isinstance(value, dict):
        return {to_snake(k): normalize(v) for k, v in value.items()}
    if isinstance(value, list):
        return [normalize(v) for v in value]
    return enum_value(value)


def _values_equal(want: Any, got: Any) -> bool:
    """Compare a contract value against an observed value. Numbers compare
    numerically (int vs float agnostic); None matches None; containers recurse."""
    if want is None:
        return got is None
    if isinstance(want, bool) or isinstance(got, bool):
        return want == got
    if isinstance(want, (int, float)) and isinstance(got, (int, float)):
        return float(want) == float(got)
    if isinstance(want, dict) and isinstance(got, dict):
        if set(want.keys()) != set(got.keys()):
            return False
        return all(_values_equal(want[k], got[k]) for k in want)
    if isinstance(want, list) and isinstance(got, list):
        return len(want) == len(got) and all(
            _values_equal(w, g) for w, g in zip(want, got)
        )
    return want == got


def assert_received(case: dict[str, Any], got: dict[str, Any] | None) -> None:
    expected = case["handler"].get("expect_received")
    if expected is None:
        return
    if got is None:
        raise DriverError(f"[{case['name']}] handler recorded no args")
    for key, want in expected.items():
        if key not in got:
            raise DriverError(
                f"[{case['name']}] handler did not receive arg {key!r}"
            )
        if not _values_equal(want, got[key]):
            raise DriverError(
                f"[{case['name']}] handler arg {key!r}: got {got[key]!r}, "
                f"want {want!r}"
            )


def _result_to_plain(result: Any) -> Any:
    """Convert the client's typed result (pydantic model or list of models) to
    plain JSON-able data with snake_case keys."""
    if isinstance(result, list):
        return [item.model_dump(mode="json") for item in result]
    return result.model_dump(mode="json")


def assert_ok(case: dict[str, Any], result: Any) -> None:
    # Normalize BOTH sides identically. `normalize` snake_cases every dict key
    # — required for struct fields (contract camelCase → generator snake_case),
    # but it can't tell a struct payload from a Map<String, String> data field,
    # so it would also rewrite map keys. Applying it to `got` as well makes any
    # such rewrite symmetric: a map key like "envName" becomes "env_name" on
    # both sides and still compares equal, so the Map round-trip is validated
    # correctly regardless of the keys' case convention.
    want = normalize(case["expect_client"]["ok"])
    got = normalize(_result_to_plain(result))
    if not _values_equal(want, got):
        raise DriverError(
            f"[{case['name']}] client result mismatch:\n"
            f" got: {json.dumps(got)}\n"
            f"want: {json.dumps(want)}"
        )


def assert_download(case: dict[str, Any], result: Any) -> None:
    """Compare the bytes the client read off a binary response (decoded as
    UTF-8) against expect_client.expect_download."""
    want = case["expect_client"]["expect_download"]
    if not isinstance(result, (bytes, bytearray)):
        raise DriverError(
            f"[{case['name']}] expected bytes from client, got "
            f"{type(result).__name__}"
        )
    got = result.decode()
    if got != want:
        raise DriverError(
            f"[{case['name']}] download mismatch: got {got!r}, want {want!r}"
        )


def assert_ok_headers(case: dict[str, Any], result: Any) -> None:
    """Compare the response-header fields the client read off the typed envelope
    against expect_client.ok_headers. The contract uses camelCase header keys
    (ratelimitRemaining, etag); the generated Python envelope exposes snake_case
    attributes (ratelimit_remaining, etag). ratelimitRemaining is a required int
    compared numerically; etag is an optional string that must equal the expected
    value or be None when the contract value is JSON null."""
    want = case["expect_client"].get("ok_headers")
    if not want:
        return
    for key, expected in want.items():
        if key == "ratelimitRemaining":
            got = result.ratelimit_remaining
        elif key == "etag":
            got = result.etag
        else:
            raise DriverError(f"[{case['name']}] unknown ok_header {key!r}")
        if not _values_equal(expected, got):
            raise DriverError(
                f"[{case['name']}] ok_header {key!r}: got {got!r}, "
                f"want {expected!r}"
            )


def assert_multi_status(case: dict[str, Any], result: Any) -> None:
    """Multi-status endpoints return an <Endpoint>Response envelope carrying the
    handler-chosen status plus an optional body. Assert the client observed the
    expected status (expect_client.status) and either an absent body
    (expect_client.ok_absent) or a body matching expect_client.ok (compared via
    the same camelCase→snake_case normalization as assert_ok)."""
    expect = case["expect_client"]
    want_status = expect["status"]
    got_status = result.status
    if got_status != want_status:
        raise DriverError(
            f"[{case['name']}] envelope status: got {got_status}, "
            f"want {want_status}"
        )
    # An ALL-TYPELESS envelope (e.g. RequeuePostResponse) has no `body`
    # attribute at all — getattr treats that the same as an absent body.
    result_body = getattr(result, "body", None)
    if expect.get("ok_absent"):
        if result_body is not None:
            raise DriverError(
                f"[{case['name']}] expected absent body, got {result_body!r}"
            )
        return
    assert_ok(case, result_body)


def assert_error_status(case: dict[str, Any], err: Exception | None) -> None:
    if err is None:
        raise DriverError(f"[{case['name']}] expected an error, got success")
    if not isinstance(err, httpx.HTTPStatusError):
        raise DriverError(
            f"[{case['name']}] expected httpx.HTTPStatusError, got "
            f"{type(err).__name__}: {err}"
        )
    expect_error = case["expect_client"].get("error")
    if expect_error is None:
        raise DriverError(f"[{case['name']}] case has no expect_client.error")
    want_status = expect_error["status_per_target"].get(TARGET)
    if want_status is None:
        raise DriverError(
            f"[{case['name']}] no status_per_target[{TARGET!r}]"
        )
    got_status = err.response.status_code
    if got_status != want_status:
        raise DriverError(
            f"[{case['name']}] error status: got {got_status}, "
            f"want {want_status}"
        )

File: crates/phoenix-codegen/tests/roundtrip/typescript/harness.ts
// Schema-agnostic round-trip boilerplate shared with the per-schema driver.ts.
//
// This module holds the parts of the TypeScript behavioral round-trip harness
// that do NOT depend on a particular generated schema: the contract.json case
// shape (`ContractCase` and friends), the `target` constant, and the generic
// assertions/helpers that compare decoded args, error statuses, etc. The
// schema-COUPLED pieces (the `Handlers` stub, the `api.*` client invocation, the
// express/createRouter server mount, and assertions that name generated types)
// live next to it in driver.ts, which imports what it needs from here.

import assert from "node:assert/strict";

import { ApiError } from "./generated/client";

export const target = "typescript" as const;

// ── contract.json schema (mirror of the language-agnostic format) ───────────

export interface ContractCase {
  name: string;
  endpoint: string;
  kind: "ok" | "error" | "constraint";
  call: {
    path_params?: Record<string, string>;
    query?: Record<string, unknown>;
    body?: Record<string, unknown>;
    headers?: Record<string, unknown>;
    // multipart/file-upload body (mutually exclusive with `body`): each file
    // part carries a UTF-8 `content` string the driver wraps in a Blob, plus
    // scalar form `fields`. The generated client builds its own FormData from
    // the typed UploadAvatarBody we construct from this.
    multipart?: {
      files: Record<string, { filename: string; content: string }>;
      fields: Record<string, unknown>;
    };
  };
  // When present, the driver answers every request itself with this canned
  // status (+ optional JSON body) INSTEAD of mounting the generated server —
  // the only way to put a status on the wire that the generated server's own
  // guard refuses (e.g. an undeclared 2xx for the client-leniency cases). The
  // stub handler is never invoked, so the ok-case hit assertion is skipped.
  raw_response?: { status: number; body?: unknown };
  handler: {
    expect_received?: Record<string, unknown>;
    returns?: unknown;
    returns_headers?: Record<string, unknown>;
    // binary download: the UTF-8 content the stub streams back as the raw
    // (non-JSON) response body (returned from the handler as a Buffer).
    returns_file?: string;
    // multi-status: the HTTP status the stub sets on the returned
    // <Endpoint>Response envelope; `returns` (when present) is the body.
    returns_status?: number;
    raises?: string;
    expect_not_called?: boolean;
  };
  expect_client: {
    ok?: unknown;
    ok_headers?: Record<string, unknown>;
    // multi-status: the status the client must observe on the envelope, and
    // (for a bodyless status like 204) that the envelope body is absent.
    status?: number;
    ok_absent?: boolean;
    // binary download: the UTF-8 content the client must read off the Blob
    // response body (present instead of `ok`).
    expect_download?: string;
    error?: {
      variant: string;
      status_per_target: Record<string, number>;
    };
  };
}

// State shared between the stub and the per-case assertions. `hit` records
// whether the relevant handler ran; `received` is the flattened map of decoded
// args the handler saw (compared against expect_received).
export interface CaseState {
  hit: boolean;
  received: Record<string, unknown>;
}

// errOrReturn signals the handler's response: throws an Error carrying the
// variant name (so the server maps it to a status) or returns the canned value.
export function signal<T>(c: ContractCase): T {
  if (c.handler.raises !== undefined) {
    throw new Error(c.handler.raises);
  }
  return c.handler.returns as T;
}

// ── assertions ──────────────────────────────────────────────────────────────

// assertReceived checks every key in expect_received against the decoded args
// the handler actually saw. Numbers compare numerically; only listed keys are
// checked. `null` in the contract means the optional arg was absent.
// UNOBSERVABLE_KEYS: keys present in the contract's expect_received that this
// target structurally cannot observe, so the driver skips their sub-assertion.
// `avatar_filename`: the generated TS server delivers multipart files as
// `Record<string,Blob>`, and a Blob carries no filename — unlike Go's
// multipart.FileHeader, the original filename is lost before it reaches the
// handler. The README explicitly permits a driver to omit checking a key it
// cannot observe ("only listed keys are checked"); the content + caption +
// thumbnail-absent assertions still fully exercise the multipart round-trip.
const UNOBSERVABLE_KEYS = new Set<string>([
  "avatar_filename",
  "thumbnail_filename",
]);

export function assertReceived(c: ContractCase, got: Record<string, unknown>): void {
  const want = c.handler.expect_received;
  if (want === undefined) return;
  for (const [k, w] of Object.entries(want)) {
    if (UNOBSERVABLE_KEYS.has(k)) continue;
    assert.ok(k in got, `[${c.name}] handler did not receive arg ${k}`);
    assert.deepStrictEqual(
      got[k],
      w,
      `[${c.name}] handler arg ${k}: got ${JSON.stringify(got[k])}, want ${JSON.stringify(w)}`,
    );
  }
}

// assertErrorStatus asserts the thrown error is an ApiError whose .status (and
// .code) match the expected per-target values.
export function assertErrorStatus(c: ContractCase, thrown: unknown): void {
  const expect = c.expect_client.error;
  assert.ok(expect !== undefined, `[${c.name}] case has no expect_client.error`);
  const wantStatus = expect.status_per_target[target];
  assert.ok(
    wantStatus !== undefined,
    `[${c.name}] no status_per_target.${target}`,
  );
  assert.ok(
    thrown instanceof ApiError,
    `[${c.name}] expected ApiError, got: ${String(thrown)}`,
  );
  assert.equal(
    thrown.status,
    wantStatus,
    `[${c.name}] error status: got ${String(thrown.status)}, want ${String(wantStatus)}`,
  );
  assert.equal(
    thrown.code,
    expect.variant,
    `[${c.name}] error code: got ${thrown.code}, want ${expect.variant}`,
  );
}
